src/meson.build

```diff
@@ -0,0 +1,424 @@
+
+tinc_dep = []
+tincd_dep = [
+  cc.find_library('m', required: false),
+]
+sptps_speed_dep = []
+sptps_test_dep = []
+sptps_keypair_dep = []
+
+if os_name == 'linux'
+  subdir('linux')
+elif os_name.endswith('bsd') or os_name in ['dragonfly', 'darwin']
+  subdir('bsd')
+elif os_name == 'sunos'
+  subdir('solaris')
+elif os_name == 'windows'
+  subdir('mingw')
+endif
+
+foreach h : check_headers
+  if cc.has_header(h)
+    cdata.set('HAVE_' + h.to_upper().underscorify(),
+              1,
+              description: '#include <' + h + '>')
+  endif
+endforeach
+
+foreach f : check_functions
+  if cc.has_function(f)
+    cdata.set('HAVE_' + f.to_upper(),
+              1,
+              description: 'function ' + f)
+  endif
+endforeach
+
+if cc.has_function('res_init', prefix: '''
+  #include <netinet/in.h>
+  #include <resolv.h>
+''')
+  cdata.set('HAVE_DECL_RES_INIT', 1)
+endif
+
+prefix_have = '''
+#include "@0@/src/config.h"
+#include "have.h"
+'''.format(build_root)
+
+type_conf = configuration_data()
+type_conf.merge_from(cdata)
+configure_file(output: 'config.h', configuration: type_conf)
+
+foreach type : check_types
+  if cc.has_type(type,
+                 prefix: prefix_have,
+                 args: cc_flags,
+                 include_directories: conf_inc)
+    name = 'HAVE_' + type.to_upper().underscorify()
+    cdata.set(name, 1, description: type)
+  endif
+endforeach
+
+dep_resolv = cc.find_library('resolv', required: false)
+
+if not cdata.has('HAVE_GETOPT_H') or not cc.has_function('getopt_long')
+  src = ['getopt.c', 'getopt1.c']
+  tinc_src += src
+  tincd_src += src
+  sptps_test_src += src
+  sptps_keypair_src += src
+endif
+
+dep_miniupnpc = dependency('miniupnpc', required: get_option('miniupnpc'))
+if dep_miniupnpc.found()
+  tincd_src += 'upnp.c'
+  tincd_dep += [
+    dependency('threads'),
+    dep_miniupnpc,
+  ]
+  cdata.set('HAVE_MINIUPNPC', 1)
+endif
+
+# The meta-dependency covers more alternatives, but is only available in 0.54+
+curses_name = meson.version().version_compare('>=0.54') ? 'curses' : 'ncurses'
+dep_curses = dependency(curses_name, required: get_option('curses'))
+if dep_curses.found()
+  cdata.set('HAVE_CURSES', 1)
+  tinc_dep += dep_curses
+endif
+
+# Some distributions do not supply pkg-config files for readline
+opt_readline = get_option('readline')
+if not opt_readline.disabled()
+  dep_readline = dependency('readline', required: false)
+  if not dep_readline.found()
+    dep_readline = cc.find_library('readline', required: opt_readline)
+  endif
+  if dep_readline.found() and cc.has_header('readline/readline.h', dependencies: dep_readline) and cc.has_header('readline/history.h', dependencies: dep_readline)
+    cdata.set('HAVE_READLINE', 1)
+    tinc_dep += dep_readline
+  endif
+endif
+
+dep_zlib = dependency('zlib', required: get_option('zlib'))
+if dep_zlib.found() and cc.has_function('compress2', dependencies: dep_zlib)
+  cdata.set('HAVE_ZLIB', 1)
+  tincd_dep += dep_zlib
+endif
```

Comment:
Hm, I wonder if this is the proper way to do it. First of all, I think you can just unconditionally add the dependency object to `tincd_dep`; if it's not required then it should be a no-op. Also, meson is quite a bit smarter than autotools when it comes to checking dependencies; I don't think we need to check for `compress2()`. At most we should specify a minimum version for the dependency.

Reply:
Even the version check is probably pretty [pointless](https://zlib.net/ChangeLog.txt):

    Changes in 1.0.8 (27 Jan 1998)
    …
    - added compress2() to allow setting the compression level